File: trees/self-balancing/simplified/RedBlackTree.java
enum Cor {
    VERMELHO, PRETO
}

class No {
    int valor;
    Cor cor;
    No esquerda, direita;
    
    public No(int valor) {
        this.valor = valor;
        this.cor = Cor.VERMELHO; // Todo novo nó é vermelho
    }
}

public class RedBlackTree {
    private No raiz;
    
    // ==================== MÉTODOS AUXILIARES ====================
    
    private boolean ehVermelho(No no) {
        return no != null && no.cor == Cor.VERMELHO;
    }
    
    // ==================== MÉTODOS DE ROTAÇÃO ====================
    
    private No rotacaoEsquerda(No pivo) {
        No filhoDireita = pivo.direita;
        pivo.direita = filhoDireita.esquerda;
        filhoDireita.esquerda = pivo;
        filhoDireita.cor = pivo.cor;
        pivo.cor = Cor.VERMELHO;
        return filhoDireita;
    }
    
    private No rotacaoDireita(No pivo) {
        No filhoEsquerda = pivo.esquerda;
        pivo.esquerda = filhoEsquerda.direita;
        filhoEsquerda.direita = pivo;
        filhoEsquerda.cor = pivo.cor;
        pivo.cor = Cor.VERMELHO;
        return filhoEsquerda;
    }
    
    // ==================== INSERÇÃO ====================
    
    public void inserir(int valor) {
        raiz = inserirRecursivo(raiz, valor);
        raiz.cor = Cor.PRETO;
    }
    
    private No inserirRecursivo(No no, int valor) {
        if (no == null) {
            return new No(valor);
        }
        
        if (valor < no.valor) {
            no.esquerda = inserirRecursivo(no.esquerda, valor);
        } else if (valor > no.valor) {
            no.direita = inserirRecursivo(no.direita, valor);
        } else {
            return no;
        }
        
        return balancear(no);
    }
    
    private No balancear(No no) {
        // Caso esquerda-direita: requer rotação dupla
        if (ehVermelho(no.esquerda) && ehVermelho(no.esquerda.direita)) {
            no.esquerda = rotacaoEsquerda(no.esquerda);
        }
        
        // Caso direita-esquerda: requer rotação dupla
        if (ehVermelho(no.direita) && ehVermelho(no.direita.esquerda)) {
            no.direita = rotacaoDireita(no.direita);
        }
        
        // Caso esquerda-esquerda
        if (ehVermelho(no.esquerda) && ehVermelho(no.esquerda.esquerda)) {
            no = rotacaoDireita(no);
        }
        
        // Caso direita-direita
        if (ehVermelho(no.direita) && ehVermelho(no.direita.direita)) {
            no = rotacaoEsquerda(no);
        }
        
        // Split de nó 4 (dois filhos vermelhos)
        if (ehVermelho(no.esquerda) && ehVermelho(no.direita)) {
            inverterCores(no);
        }
        
        return no;
    }
    
    private void inverterCores(No no) {
        no.cor = Cor.VERMELHO;
        if (no.esquerda != null) no.esquerda.cor = Cor.PRETO;
        if (no.direita != null) no.direita.cor = Cor.PRETO;
    }
    
    // ==================== OPERAÇÕES DE BUSCA ====================
    
    public No buscar(int valor) {
        No atual = raiz;
        while (atual != null && atual.valor != valor) {
            atual = (valor < atual.valor) ? atual.esquerda : atual.direita;
        }
        return atual;
    }
    
    public No encontrarMinimo() {
        if (raiz == null) return null;
        No atual = raiz;
        while (atual.esquerda != null) {
            atual = atual.esquerda;
        }
        return atual;
    }
    
    public No encontrarMaximo() {
        if (raiz == null) return null;
        No atual = raiz;
        while (atual.direita != null) {
            atual = atual.direita;
        }
        return atual;
    }
    
    // ==================== PERCURSO ====================
    
    public void emOrdem(No no) {
        if (no != null) {
            emOrdem(no.esquerda);
            System.out.print(no.valor + "(" + 
                (no.cor == Cor.VERMELHO ? "V" : "P") + ") ");
            emOrdem(no.direita);
        }
    }
    
    public No getRaiz() {
        return raiz;
    }
    
    // ==================== MÉTODO MAIN ====================
    
    public static void main(String[] args) {
        RedBlackTree arvore = new RedBlackTree();
        
        int[] valores = {10, 20, 30, 15, 25, 5, 1};
        System.out.println("=== INSERÇÃO ===");
        for (int valor : valores) {
            arvore.inserir(valor);
            System.out.print("Após inserir " + valor + ": ");
            arvore.emOrdem(arvore.getRaiz());
            System.out.println();
        }
        
        System.out.println("\n=== BUSCA ===");
        No encontrado = arvore.buscar(15);
        if (encontrado != null) {
            System.out.println("Encontrado: " + encontrado.valor);
        }
        
        System.out.println("\n=== MIN/MAX ===");
        No noMinimo = arvore.encontrarMinimo();
        No noMaximo = arvore.encontrarMaximo();
        System.out.println("Mínimo: " + noMinimo.valor + ", Máximo: " + noMaximo.valor);
    }
}

File: trees/self-balancing/simplified/RedBlackTree.c
#include <stdio.h>
#include <stdlib.h>

typedef enum {
    VERMELHO,
    PRETO
} Cor;

typedef struct No {
    int valor;
    Cor cor;
    struct No *esquerda, *direita;
} No;

typedef struct {
    No *raiz;
} ArvoreRubroNegra;

// ==================== MÉTODOS AUXILIARES ====================

int ehVermelho(No *no) {
    return no != NULL && no->cor == VERMELHO;
}

No* criarNo(int valor) {
    No *novo = (No*)malloc(sizeof(No));
    novo->valor = valor;
    novo->cor = VERMELHO; 
    novo->esquerda = NULL;
    novo->direita = NULL;
    return novo;
}

// ==================== MÉTODOS DE ROTAÇÃO ====================

No* rotacaoEsquerda(No *pivo) {
    No *filhoDireita = pivo->direita;
    pivo->direita = filhoDireita->esquerda;
    filhoDireita->esquerda = pivo;
    filhoDireita->cor = pivo->cor;
    pivo->cor = VERMELHO;
    return filhoDireita;
}

No* rotacaoDireita(No *pivo) {
    No *filhoEsquerda = pivo->esquerda;
    pivo->esquerda = filhoEsquerda->direita;
    filhoEsquerda->direita = pivo;
    filhoEsquerda->cor = pivo->cor;
    pivo->cor = VERMELHO;
    return filhoEsquerda;
}

// ==================== BALANCEAMENTO ====================

void inverterCores(No *no) {
    no->cor = VERMELHO;
    if (no->esquerda != NULL) no->esquerda->cor = PRETO;
    if (no->direita != NULL) no->direita->cor = PRETO;
}

No* balancear(No *no) {
    if (ehVermelho(no->esquerda) && ehVermelho(no->direita)) {
        inverterCores(no);
    }
    
    if (ehVermelho(no->direita) && !ehVermelho(no->esquerda)) {
        no = rotacaoEsquerda(no);
    }
    
    if (ehVermelho(no->esquerda) && ehVermelho(no->esquerda->esquerda)) {
        no = rotacaoDireita(no);
    }
    
    if (ehVermelho(no->esquerda) && ehVermelho(no->direita)) {
        inverterCores(no);
    }
    
    return no;
}

// ==================== INSERÇÃO ====================

No* inserirRecursivo(No *no, int valor) {
    if (no == NULL) {
        return criarNo(valor);
    }
    
    if (valor < no->valor) {
        no->esquerda = inserirRecursivo(no->esquerda, valor);
    } else if (valor > no->valor) {
        no->direita = inserirRecursivo(no->direita, valor);
    } else {
        return no;
    }
    
    return balancear(no);
}

void inserir(ArvoreRubroNegra *arvore, int valor) {
    arvore->raiz = inserirRecursivo(arvore->raiz, valor);
    arvore->raiz->cor = PRETO; 
}

// ==================== OPERAÇÕES DE BUSCA ====================

No* buscar(ArvoreRubroNegra *arvore, int valor) {
    No *atual = arvore->raiz;
    while (atual != NULL && atual->valor != valor) {
        atual = (valor < atual->valor) ? atual->esquerda : atual->direita;
    }
    return atual;
}

No* encontrarMinimo(ArvoreRubroNegra *arvore) {
    if (arvore->raiz == NULL) return NULL;
    No *atual = arvore->raiz;
    while (atual->esquerda != NULL) {
        atual = atual->esquerda;
    }
    return atual;
}

No* encontrarMaximo(ArvoreRubroNegra *arvore) {
    if (arvore->raiz == NULL) return NULL;
    No *atual = arvore->raiz;
    while (atual->direita != NULL) {
        atual = atual->direita;
    }
    return atual;
}

// ==================== PERCURSO ====================

void emOrdem(No *no) {
    if (no != NULL) {
        emOrdem(no->esquerda);
        printf("%d(%c) ", no->valor, no->cor == VERMELHO ? 'V' : 'P');
        emOrdem(no->direita);
    }
}

// ==================== LIBERAÇÃO DE MEMÓRIA ====================

void liberarArvore(No *no) {
    if (no != NULL) {
        liberarArvore(no->esquerda);
        liberarArvore(no->direita);
        free(no);
    }
}

// ==================== FUNÇÃO MAIN ====================

int main() {
    ArvoreRubroNegra arvore;
    arvore.raiz = NULL;
    
    int valores[] = {10, 20, 30, 15, 25, 5, 1};
    int n = sizeof(valores) / sizeof(valores[0]);
    
    printf("=== INSERCAO ===\n");
    for (int i = 0; i < n; i++) {
        inserir(&arvore, valores[i]);
        printf("Apos inserir %d: ", valores[i]);
        emOrdem(arvore.raiz);
        printf("\n");
    }
    
    printf("\n=== BUSCA ===\n");
    No *encontrado = buscar(&arvore, 15);
    if (encontrado != NULL) {
        printf("Encontrado: %d\n", encontrado->valor);
    }
    
    printf("\n=== MIN/MAX ===\n");
    No *noMinimo = encontrarMinimo(&arvore);
    No *noMaximo = encontrarMaximo(&arvore);
    printf("Minimo: %d, Maximo: %d\n", noMinimo->valor, noMaximo->valor);
    
    liberarArvore(arvore.raiz);
    
    return 0;
}
